pkg/metrics/metrics.go

```diff
@@ -0,0 +1,100 @@
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Error to be  returned when marketplace is unable to reach quay.io.
+type OpsrcError string
+
+const (
+	// Default values which are to be used as labels
+	// for prometheus metrics.
+	Error     = "opsrc_error"
+	OpsrcName = "opsrc_name"
+
+	// MetricsPath, MetricsPort and MetricsHost where custom metrics would be exposed.
+	MetricsPath = "/metrics"
+	MetricsPort = 8383
+	MetricsHost = "0.0.0.0"
+
+	// UnreachanbeError is the error message returned when quay.io is down.
+	UnreachableError OpsrcError = "quay_unavailabe"
+
+	// NoHostExsistsError is the error message returned when host address is wrong.
+	NoHostExsistsError OpsrcError = "no_host_exsists"
+)
+
+// Metric variables that report error based on the operator-source
+// which fails to reconcile.
+var (
+	operatorSourceError = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "operator_source_failure_count",
+			Help: "Monotonic count of times operator-source could not download packages",
+		},
+		[]string{OpsrcName, Error},
+	)
+	// List of custom metrics which are to be collected.
+	metricsList = []prometheus.Collector{
+		operatorSourceError,
+	}
+)
+
+// registerMetrics registers the metrics with prometheus.
+func registerMetrics() error {
+	for _, metric := range metricsList {
+		err := prometheus.Register(metric)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RecordMetrics starts the server, records the custom metrics and exposes
+// the metrics at the specified endpoint.
+func RecordMetrics() {
+	// Register metrics for the operator with the prometheus.
+	registerMetrics()
+
+	// Start the server and expose the registered metrics.
+	http.Handle(MetricsPath, promhttp.Handler())
+	//port := ":" + MetricsPort
+	port := fmt.Sprintf(":%d", MetricsPort)
+	go http.ListenAndServe(port, nil)
+}
+
+// counterForOpsrcError returns the number of times error was encountered while
+// contacting quay for operstor-sources.
+func counterForOpsrcError(opsrc, opsrcError string) prometheus.Counter {
+	return operatorSourceError.WithLabelValues(opsrc, opsrcError)
+}
+
+// updateMetrics takes the name of the operator source which fails to reconcile, and
+// increments the error count of the respective operator metric.
+func UpdateMetrics(opsrc, opsrcError string) {
```

Comment:
I think this should be `func UpdateMetrics(opsrc, opsrcError OpsrcError)`?

Reply:
The labels of the prometheus metrics only take in String type. So I created a function `OpsrcErrorToString`, which would just cast it into a string only while updating the labels.

Reply:
I had used that in `configuring.go`, while incrementing metrics. But that doesn't serve the purpose. So I have updated it here! :+1: